### PR TITLE
wave 2/3 requirements met, not working

### PR DESCRIPTION
**Pull request by Abiaina**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Question 	|   Response	|
|---	|---	|
|   Why is it useful to put classes inside modules? 	|   So I can reuse specific methods and have class variables to store info across multiple instances of a class.  allows for dry code	|
|   What is accomplished with `raise ArgumentError`? 	|   to let user know that they entered invalid parameters 	|
|   Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods?	|   because we want to look at data across all instances within a class or related classes	|
|   Why does it make sense to use inheritance for the online order?  |  because both order and online orders are orders of the module grocery.|
|   Did the presence of automated tests change the way you thought about the problem? How? | Yes, it made me run testing more often. But when i had to write my own tests I had a very hard time following the tdd paradigm .|

### CODE REVIEW

**Comment by Abiaina** [comment]
updated code

File: specs/order_spec.rb [hunk 1]
```diff
@@ -78,33 +80,44 @@
...
   describe "Order.find" do
     it "Can find the first order from the CSV" do
-      # TODO: Your test code here!
+      result = Grocery::Order.find(1)
+      result.must_equal ([1, {"Slivered Almonds":22.88, "Wholewheat flour":1.93, "Grape Seed Oil":74.9}])
     end
 
     it "Can find the last order from the CSV" do
-      # TODO: Your test code here!
+      result = Grocery::Order.find(100)
```
Comment (CheezItMan) [inline specs/order_spec.rb:135]: Test not passing, `find` should return an `Order` and not a 2-element array with a number and a hash.

So I would do
```ruby
result.id.must_equal 100
result.products["Allspice"].must_equal 64.74
```

File: lib/order.rb [hunk 2]
```diff
@@ -1,18 +1,80 @@
...
+        subtotal += price
+      end
+      tax = subtotal * 0.075
+      @gross_total = (subtotal + tax).round(2)
+      return @gross_total
+    end
+    # Takes user input product, price and order id
+    # Allows user to add new order, or add product and its price to existing order
+    # Appends new order onto existing orders csv file
+    def add_product (product_name, product_price, add_id)
```
Comment (CheezItMan) [inline lib/order.rb]: This method should be getting a product name & price, **not** an id.

This method should only:
- Check to see if a product with that name exists in `@products`
  - If it does then return `false`
  - if it does not then add that name/price combo to the `products` hash.

You have this method doing **waaaay** too much.  Read the tests and do **only** what they require.

File: lib/order.rb [hunk 3]
```diff
@@ -1,18 +1,80 @@
+require 'awesome_print'
+require 'csv'
+
 module Grocery
   class Order
-    attr_reader :id, :products
+    attr_reader :id
+    attr_accessor :order, :products, :gross_total
```
Comment (CheezItMan) [inline lib/order.rb]: Why an `attr_accessor` for `products` and why have an `order` accessor?

File: lib/order.rb [hunk 4]
```diff
@@ -1,18 +1,80 @@
...
+
+    @@organized_orders = []
```
Comment (CheezItMan) [inline lib/order.rb]: While Class variables can be used, it's tricky and easier to do this without'em.

File: lib/order.rb [hunk 5]
```diff
@@ -1,18 +1,80 @@
...
 
-    def initialize(id, products)
+    def initialize (id, products)
```
Comment (CheezItMan) [inline lib/order.rb]: Don't put a space between the method name and the (

File: lib/order.rb [hunk 6]
```diff
@@ -1,18 +1,80 @@
...
 
     def total
-      # TODO: implement total
+      subtotal = 0
+      @products.each do |product, price|
+        subtotal += price
+      end
+      tax = subtotal * 0.075
+      @gross_total = (subtotal + tax).round(2)
+      return @gross_total
```
Comment (CheezItMan) [inline lib/order.rb]: Well done, but `@gross_total` doesn't need to be an instance variable, it can be local.

File: lib/order.rb [hunk 7]
```diff
@@ -1,18 +1,80 @@
...
+    end
+    # Takes user input product, price and order id
+    # Allows user to add new order, or add product and its price to existing order
+    # Appends new order onto existing orders csv file
+    def add_product (product_name, product_price, add_id)
+      if add_id != "new"
```
Comment (CheezItMan) [inline lib/order.rb]: Why this conditional?

File: lib/order.rb [hunk 8]
```diff
@@ -1,18 +1,80 @@
...
+        add_id = add_id.to_i
+        add_products = self.find_by_id(add_id)
```
Comment (CheezItMan) [inline lib/order.rb]: Adding a product to **this** order doesn't require you to **find** the order.

File: lib/order.rb [hunk 9]
```diff
@@ -1,18 +1,80 @@
...
+            else
+              add_products.store(product_name, product_price)
+              puts "#{product_name} has been added to #{add_id}"
+              return true
+            end
+          end
+        else
+          new_id = orders.length + 1
+          new_products = {product_name => product_price}
+          CSV.open("../support/orders.csv", 'a') do |orders_csv|
```
Comment (CheezItMan) [inline lib/order.rb]: Adding a product should also not require reading the CSV.

File: lib/order.rb [hunk 10]
```diff
@@ -1,18 +1,80 @@
...
+        if @id == find_id
+          found_order = order.products
+        end
+      end
+      return found_order
     end
 
-    def add_product(product_name, product_price)
-      # TODO: implement add_product
+    def self.all
```
Comment (CheezItMan) [inline lib/order.rb]: This should **mostly** work, but if `self.all` is called multiple times then `@@organized_orders` will get longer and longer.  You will either need to empty out `@@organized_orders` each time `self.all` is called or check to see if it's not nil and if so, then don't read the file and instead just return `@@organized_orders`.

In general, you should almost **never** use class variables.  They can provide a modicum of efficiency in this case, but make coding the methods much tricker.

File: specs/order_spec.rb [hunk 11]
```diff
@@ -78,33 +80,44 @@
...
 end
 
-# TODO: change 'xdescribe' to 'describe' to run these tests
-xdescribe "Order Wave 2" do
+describe "Order Wave 2" do
   describe "Order.all" do
     it "Returns an array of all orders" do
-      # TODO: Your test code here!
+      result = Grocery::Order.all
+      result.must_equal (CSV.read("/support/orders.csv"))
```
Comment (CheezItMan) [inline specs/order_spec.rb]: `Order.all` should return an array of `Order` instances **not** an array of arrays!

File: specs/order_spec.rb [hunk 12]
```diff
@@ -78,33 +80,44 @@
...
     end
 
     it "Returns accurate information about the first order" do
-      # TODO: Your test code here!
+      result = Grocery::Order.all
+      result_first = result[0]
+      test_result = CSV.read("/support/orders.csv")
+      test_first = test_result[0]
+      result_first.must_equal test_first
```
Comment (CheezItMan) [inline specs/order_spec.rb]: Again you're not creating an Order instance , or testing to see if the `id` field matches the ID from the CSV.

File: specs/order_spec.rb [hunk 13]
```diff
@@ -78,33 +80,44 @@
...
     end
 
     it "Returns accurate information about the last order" do
-      # TODO: Your test code here!
+      result = Grocery::Order.all
+      result_first = result[-1]
```
Comment (CheezItMan) [inline specs/order_spec.rb]: Don't call the variable `result_first` if it's the **last** element.

File: specs/order_spec.rb [hunk 14]
```diff
@@ -78,33 +80,44 @@
...
+      test_result = CSV.read("/support/orders.csv")
+      test_first = test_result[-1]
+      result_first.must_equal test_first
```
Comment (CheezItMan) [inline specs/order_spec.rb]: Same issue with your first order test.

File: specs/order_spec.rb [hunk 15]
```diff
@@ -78,33 +80,44 @@
...
     end
   end
 
   describe "Order.find" do
     it "Can find the first order from the CSV" do
-      # TODO: Your test code here!
+      result = Grocery::Order.find(1)
+      result.must_equal ([1, {"Slivered Almonds":22.88, "Wholewheat flour":1.93, "Grape Seed Oil":74.9}])
```
Comment (CheezItMan) [inline specs/order_spec.rb]: Test not passing, `find` should return an `Order` and not a 2-element array with a number and a hash.

So I would do
```ruby
result.id.must_equal 1
result.products["Slivered Almonds"].must_equal 22.88
```

File: specs/online_order_spec.rb [hunk 16]
```diff
@@ -1,90 +1,117 @@
 require 'minitest/autorun'
```
Comment (CheezItMan) [inline specs/online_order_spec.rb:1]: For this file you've got a lot of similar problems with `order_spec.rb`.  I think fixing these would lead to you  fixing the others.

File: lib/order.rb [hunk 17]
```diff
@@ -1,18 +1,80 @@
...
+          CSV.open("../support/orders.csv", 'a') do |orders_csv|
+            orders_csv << ["#{new_id}", "#{product_name}:#{product_price}"]
+          end
+          # return is better because allows for continuous updating of csv from multiple programs.
+          @@organized_orders << Order.new(new_id, new_products)
+        end
+      end
+    end
+
+    def self.find (find_id)
```
Comment (CheezItMan) [inline lib/order.rb]: This method should either test to see if `@@organized_orders` is nil and if so, then call `self.all`

File: specs/online_order_spec.rb [hunk 18]
```diff
@@ -1,90 +1,117 @@
...
-      # TODO: Your test code here!
+      online_order = OnlineOrder.new(1, 1, "paid", {"tempeh":4.99, "pokebowl":9.00})
+      test = online_order.find_by_customer(1)
+      test.must_equal ([1, 1, "paid", {"tempeh":4.99, "pokebowl":9.00}])
     end
 
     it "Can access the online order status" do
-      # TODO: Your test code here!
+      online_order = OnlineOrder.new(1, 1, "paid", {"tempeh":4.99, "pokebowl":9.00})
+      test = online_order.status
```
Comment (CheezItMan) [inline specs/online_order_spec.rb]: A better variable name would be `test_total`

File: lib/online_order.rb [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+  def add_product (product_name, product_price, add_id)
+    if @status == "pending" || @status == "paid"
+      super add_product(product_name, product_price, add_id)
+    else
+      raise ArgumentError.new "Argument Error: product status."
+    end
+  end
+
+  def self.all
+    @@organized_online_orders.clear
```
Comment (CheezItMan) [inline lib/online_order.rb]: This works, but you still have trouble in your `Order.all`

File: lib/online_order.rb [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+    @@organized_online_orders.each do |order|
+      if order.order_id == order_id
+        found_order = order.products
+      end
+    end
+    return found_order
+  end
+
+  def self.find_by_customer (customer_id_entry)
+    found_order = false
```
Comment (CheezItMan) [inline lib/online_order.rb]: You need to test and make sure that `@@organized_online_orders` is not `nil` and if it is call `OnlineOrder.all`.

File: specs/online_order_spec.rb [hunk 21]
```diff
@@ -1,90 +1,117 @@
...
+      end
+
+      it "Raises an error if the customer does not exist" do
+        result = OnlineOrder.find_by_customer(1000)
+        result.must_equal false
+      end
+
+      it "Returns an empty array if the customer has no orders" do
+        result = OnlineOrder.find_by_customer(1000)
+        result.must_equal false
```
Comment (CheezItMan) [inline specs/online_order_spec.rb]: `false` is not an empty array.  Read the test description in the `it` block

File: lib/online_order.rb [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    @customer_id = customer_id.to_i
+    @order_id = order_id.to_i
+    @status = order_status.to_sym
+    if !status_array.include?(order_status)
+      @status = :pending
+    end
+    @products = products
+  end
+
+  def calculate_online_total
```
Comment (CheezItMan) [inline lib/online_order.rb]: This should have the same name as the parent class, not `online_total`

File: lib/online_order.rb [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+      @@organized_online_orders << OnlineOrder.new(order[0].to_i, order[2], order[3].to_sym, products)
+    end
+    return @@organized_online_orders
+  end
+
+  def self.find (order_id)
+    found_order = false
+    @@organized_online_orders.each do |order|
+      if order.order_id == order_id
+        found_order = order.products
```
Comment (CheezItMan) [inline lib/online_order.rb]: This shouldn't be the order's products, but rather the order itself.

**Comment by CheezItMan** [comment]
# Grocery Store
## What We're Looking For

|  Feature  |   Feedback|
|---	|---	|
|  **Baseline** |   |
|   Answered comprehension questions	|   	|
|   Used Git Regularly	|  More granular commits would be better.  Making more commits and talking about the functionality added rather than a generic message about tests failing.  |
|  **Wave 1** |   |
|   All provided tests pass | Many tests failing with errors |
|   Using the appropriate `attr_` for instance variables | Check, although you are using `attr_accessor` for attributes that don't need to be accessors   |
|  **Wave 2** |   |
|   All stubbed tests are implemented fully and pass | **Nope** |
|   Appropriately parses the product data from CSV file in `Order.all` | Not in `Order.all` see my in-code notes  |
|   Used `CSV` library only in `Order.all` (not in `Order.find`) |  **Nope** You have it in `add_product` for some reason |
|   Used `Order.all` to get order list in `Order.find`	|  You're using a class variable, but you need to ensure that the variable is not nil, see my in-code notes.  |
|  **Wave 3** |   |
|   All stubbed tests are implemented fully and pass | **Nope**  |
|   Used inheritance in the initialize for online order |   Check |
|   Used inheritance for the total method in online order | You didn't override the parent class method by naming it the same.  |
|   Appropriately created Customer class and class methods `all` & `find` |  |
|   Use `CSV` library only in `OnlineOrder.all` | Check |
|   Used `all` to get order list in `find` | Used a class variable like in `Order.all`, similar problems. |
|   Appropriately searches for Customer orders in `find_by_customer`  | Similar problems |
|   **Additional Notes**	|   	|
|  Summary | I'd like you to review Grocery Store with a tutor or myself or Dee 1-on-1.  Specifically the `Order` class.  Your `add_product` has some issues I've highlighted in your code  In the future focus on **one** test at a time, get it to pass and then move to the next.  Don't try to get everything to work and **then** do the test.  You'll find it goes faster.  Especially as it might force you to get help early.
